key.go

```diff
@@ -0,0 +1,806 @@
+	return encMode.Marshal(int64(ko))
+}
+
+// UnmarshalCBOR populates the KeyOp from the provided CBOR value (must be int
+// or tstr).
+func (ko *KeyOp) UnmarshalCBOR(data []byte) error {
+	var raw intOrStr
+
+	if err := raw.UnmarshalCBOR(data); err != nil {
+		return fmt.Errorf("invalid key_ops value %w", err)
+	}
+
+	if raw.IsString() {
+		v := KeyOpFromString(raw.String())
+		if v == KeyOpInvalid {
+			return fmt.Errorf("unknown key_ops value %q", raw.String())
+		}
+
+		*ko = v
+	} else {
+		v := raw.Int()
+		*ko = KeyOp(v)
+
+		if !ko.IsSupported() {
+			return fmt.Errorf("unknown key_ops value %d", v)
+		}
+	}
+
+	return nil
+}
+
+const (
+	// Octet Key Pair
+	KeyTypeOkp KeyType = 1
+	// Elliptic Curve Keys w/ x- and y-coordinate pair
+	KeyTypeEc2 KeyType = 2
+	// Symmetric Keys
+	KeyTypeSymmetric KeyType = 4
+)
+
+type KeyType int64
+
+// KeyTypeFromString returns the KeyType corresponding to the specified name.
+func KeyTypeFromString(v string) (KeyType, error) {
+	switch v {
+	case "OKP":
+		return KeyTypeOkp, nil
+	case "EC2":
+		return KeyTypeEc2, nil
+	case "Symmetric":
+		return KeyTypeSymmetric, nil
+	default:
+		return KeyType(0), fmt.Errorf("unknown key type value %q", v)
+	}
+}
+
+func (kt KeyType) String() string {
+	switch kt {
+	case KeyTypeOkp:
+		return "OKP"
+	case KeyTypeEc2:
+		return "EC2"
+	case KeyTypeSymmetric:
+		return "Symmetric"
+	default:
+		return "unknown key type value " + strconv.Itoa(int(kt))
+	}
+}
+
+// MarshalCBOR marshals the KeyType as a CBOR int.
+func (kt KeyType) MarshalCBOR() ([]byte, error) {
+	return encMode.Marshal(int(kt))
+}
+
+// UnmarshalCBOR populates the KeyType from the provided CBOR value (must be
+// int or tstr).
+func (kt *KeyType) UnmarshalCBOR(data []byte) error {
+	var raw intOrStr
+
+	if err := raw.UnmarshalCBOR(data); err != nil {
+		return fmt.Errorf("invalid key type value: %w", err)
+	}
+
+	if raw.IsString() {
+		v, err := KeyTypeFromString(raw.String())
+
+		if err != nil {
+			return err
+		}
+
+		*kt = v
+	} else {
+		v := raw.Int()
+
+		if v == 0 {
+			// 0  is reserved, and so can never be valid
+			return fmt.Errorf("invalid key type value 0")
+		}
+
+		if v > 4 || v < 0 || v == 3 {
+			return fmt.Errorf("unknown key type value %d", v)
+		}
+
+		*kt = KeyType(v)
+	}
+
+	return nil
+}
+
+const (
+
+	// Invalid/unrecognised curve
+	CurveInvalid Curve = 0
+
+	// NIST P-256 also known as secp256r1
+	CurveP256 Curve = 1
+
+	// NIST P-384 also known as secp384r1
+	CurveP384 Curve = 2
+
+	// NIST P-521 also known as secp521r1
+	CurveP521 Curve = 3
+
+	// X25519 for use w/ ECDH only
+	CurveX25519 Curve = 4
+
+	// X448 for use w/ ECDH only
+	CurveX448 Curve = 5
+
+	// Ed25519 for use /w EdDSA only
+	CurveEd25519 Curve = 6
+
+	// Ed448 for use /w EdDSA only
+	CurveEd448 Curve = 7
+)
+
+// Curve represents the EC2/OKP key's curve. See:
+// https://datatracker.ietf.org/doc/html/rfc8152#section-13.1
+type Curve int64
+
+func CurveFromString(v string) (Curve, error) {
```

Comment:
IMO there is no need to return an error, checking for `CurveInvalid` is enough.

Reply:
That would be incositent with the over API, and returning an error allows for more nuanced error repoting (the offending value is preserved in the error message).

Reply:
Good point. Then `KeyOpFromString` should also follow the same pattern, it currently doesn't return an error.

Reply:
Good point. I'll make it private.